app/grants/models.py

```diff
@@ -1063,6 +1073,61 @@ def update_tx_status(self):
             self.success = split_tx_status == 'success'
             self.split_tx_confirmed = True
 
+    def verify_transactions(self):
+        from dashboard.utils import get_web3
+
+        def same_address(a, b):
+            return a.lower() == b.lower()
+
+        web3 = get_web3(self.subscription.network)
+
+        approve_contract = web3.eth.contract(address=None, abi=erc20_abi)
+        splitter_contract = web3.eth.contract(address=settings.SPLITTER_CONTRACT_ADDRESS, abi=splitter_abi)
+
+        try:
+            approve_tx = web3.eth.getTransaction(self.tx_id)
+            splitter_tx = web3.eth.getTransaction(self.split_tx_id)
+        except:
+            return False, 'error getting transactions'
+
+        if approve_tx is None or splitter_tx is None:
+            return False, 'transactions not found'
+
+        try:
+            approve_call = decode_function_input(approve_contract, approve_tx.input)
+            split_call = decode_function_input(splitter_contract, splitter_tx.input)
+        except:
+            return False, 'cannot decode contract calls'
+
+        if not same_address(approve_tx['from'], splitter_tx['from']):
+            return False, 'approve and split transactions sent from different addresses'
+
+        if not same_address(approve_tx['to'], self.subscription.token_address):
+            return False, 'incorrect token approved'
+
+        if not same_address(splitter_tx['to'], settings.SPLITTER_CONTRACT_ADDRESS):
+            return False, 'incorrect splitter used'
+
+        if not same_address(approve_call[1]['_spender'], splitter_tx['to']):
+            return False, 'approval is not given to splitter'
+
+        if not same_address(split_call[1]['tokenAddress'], approve_tx['to']):
+            return False, 'splitter uses incorrect token'
```

Comment:
what transactions did you use to test this against?

Reply:
Here's a successful one I've used for testing:

tx_id: https://etherscan.io/tx/0x45c9d0b6d29386d5996d58d6687bb27be13fb7488493cb6feb16db22920f1aba
split_tx_id: https://etherscan.io/tx/0x6fae97cc0e2810b3929ce53e95a0e1c4959362cce9818d8631934bcca83f96b7

Here's the one where all calls were made, but amount approved and actually sent is incorrect (this is trickiest case, because here everything looks correct, except actual amount of tokens).

tx_id: https://etherscan.io/tx/0xe83d42ed1d6305ac099adb61929af921fd8ea3ca61a6585c1c6a840d452bfe02
split_tx_id: https://etherscan.io/tx/0x0f4837fc443dea4ae50e511c2f79960675e99287d97891aa6ded06a539bf5fec

I've created contributions with these tx's using Django shell, then ran `verify_contributions` against them.

P. S. I've used some other ones, but these are the ones that are still in database.